examples/go/vm-reconfig-via-tag/handler/handler.go

```diff
@@ -0,0 +1,564 @@
+)
+
+const cfgPath = "/var/openfaas/secrets/vcconfig"
+
+var (
+	vsClt *vsClient // Persist vSphere connection.
+	once  sync.Once // For handleSignal() to be called once.
+	debug bool      // True for verbose logging.
+)
+
+// vcConfig represents the toml vcconfig file
+type vcConfig struct {
+	VCenter struct {
+		Server, User, Password string
+		Insecure               bool
+	}
+}
+
+// cloudEvent is a subsection of a Cloud Event.
+type cloudEvent struct {
+	Data types.Event `json:"data,omitempty"`
+}
+
+type vsClient struct {
+	govmomi *govmomi.Client
+	rest    *rest.Client
+	tagMgr  *tags.Manager
+}
+
+type vmConfig struct {
+	name    string
+	valBool bool
+	valInt  int
+}
+
+// Handle a function invocation
+func Handle(req handler.Request) (handler.Response, error) {
+	ctx := context.Background()
+	debug = setVerbosity()
+
+	// Load config every time, to ensure the most updated version is used.
+	cfg, err := loadConfig(cfgPath)
+	if err != nil {
+		return handlerResponseWithError("load vcconfig: %w",
+			http.StatusInternalServerError,
+			err,
+		)
+	}
+
+	// The Mananged Object Reference for the VM that powered on.
+	vmMOR, err := eventVmMOR(req.Body)
+	if err != nil {
+		return handlerResponseWithError(
+			"retrieve VM object: %w",
+			http.StatusInternalServerError,
+			err,
+		)
+	}
+
+	if vsClt == nil {
+		// Save vSphere connection in a global variable.
+		vsClt, err = newClient(ctx, cfg)
+		if err != nil {
+			return handlerResponseWithError(
+				"connect to vSphere: %w",
+				http.StatusInternalServerError,
+				err,
+			)
+		}
+	}
+
+	// Signals wil cause vSphere APIs to log out.
+	once.Do(func() {
+		go handleSignal(ctx, vsClt)
+	})
+
+	// Check REST connection by getting tag objects.
+	tagObjs, err := vsClt.attachedTags(ctx, *vmMOR)
+	if err != nil {
+		// Retry once if not connected after creating new client.
+		if strings.Contains(err.Error(), "401 Unauthorized") {
+			if debug {
+				log.Println("401 Unauthorized. Retry connection to Govmomi.")
+			}
+
+			vsClt, err = newClient(ctx, cfg)
+		}
+
+		if err == nil {
+			tagObjs, err = vsClt.attachedTags(ctx, *vmMOR)
+		}
+	}
+
+	if err != nil {
+		return handlerResponseWithError(
+			"get tags attached to VM: %w",
+			http.StatusInternalServerError,
+			err,
+		)
+	}
+
+	// Check SOAP connection by getting managed object VM.
+	moVM, err := vsClt.moVirtualMachine(ctx, *vmMOR)
+	if err != nil {
+		// Retry once after creating new client.
+		if strings.Contains(err.Error(), "NotAuthenticated") {
+			if debug {
+				log.Println("NotAuthenticated. Retry connection to VAPI.")
+			}
+
+			vsClt, err = newClient(ctx, cfg)
+		}
+
+		if err == nil {
+			moVM, err = vsClt.moVirtualMachine(ctx, *vmMOR)
+		}
+	}
+
+	if err != nil {
+		return handlerResponseWithError(
+			"retrieve current VM configs: %w",
+			http.StatusInternalServerError,
+			err,
+		)
+	}
+
+	// Get the current configurations in a list from the VM object.
+	currCfgs := currentConfigs(ctx, moVM)
```

Comment:
Nit: `getConfigs`

Reply:
I don't think `getConfigs` is the most clear. Where are we getting configs from? The tags? the VM? And if I call the function `vmConfigs` it's also not exactly right because I'm not getting all the vmConfigs, just the ones I care about.

And as mentioned before, I think `get` doesn't help clarify what the function does. It's pretty clear I'm getting current configs. I think I'll leave this name, despite not being great, but I can't think of better.

Reply:
Agree.